🤼 DisputeGame interfaces (#5304)
ethereum-optimism/optimism: specs/dispute-game-interface.md

Reviewer: I'd love to see glossary entries be included along with this document. Otherwise it feels like quite a bit of jargon for outsiders to learn.

Author: Merging this PR into Mark's - will address there.

single dispute game will not result in the bug becoming consensus.

## Types

For added context, we define a few types that are used in the following snippets.

```solidity
/// @notice The type of proof system being used.
enum GameType {
    /// @dev The game will use a `IDisputeGame` implementation that utilizes fault proofs.
    FAULT,
    /// @dev The game will use a `IDisputeGame` implementation that utilizes validity proofs.
    VALIDITY,
    /// @dev The game will use a `IDisputeGame` implementation that utilizes attestation proofs.
    ATTESTATION
}

/// @notice The current status of the dispute game.
enum GameStatus {
    /// @dev The game is currently in progress, and has not been resolved.
    IN_PROGRESS,
    /// @dev The game has concluded, and the `rootClaim` was challenged successfully.
    CHALLENGER_WINS,
    /// @dev The game has concluded, and the `rootClaim` could not be contested.
    DEFENDER_WINS
}

/// @notice A dedicated timestamp type.
type Timestamp is uint64;

/// @notice A `Claim` type represents a 32 byte hash or other unique identifier for a claim about
///         a certain piece of information.

> /// @dev For the `FAULT` `GameType`, this will be a MPT root representing the state of

Reviewer 2: Will this be a mpt root or will it be an output proposal? I think if its an mpt root then we just need the preimage of the output proposal

Author: For the fault proof game, it'll be the merklized state of the fault proof program's memory at the end of execution if I understand correctly.

///      the fault proof program at the end of the state transition.
///      For the `ATTESTATION` `GameType`, this will be an output root.
type Claim is bytes32;
```

## `DisputeGameFactory` Interface

The dispute game factory is responsible for creating new `DisputeGame` contracts
given a `GameType` and a root `Claim`. Challenger agents will listen to the
`DisputeGameCreated` events that are emitted by the factory in order to keep up
with on-going disputes in the protocol.

> For the factory, we use a [Huff](https://huff.sh) implementation of

Reviewer 2: perhaps don't use "we" and instead describe things in third person, like "For the factory, a Huff impl of ... is used"

[`clones-with-immutable-args`](https://github.com/wighawag/clones-with-immutable-args/tree/master)
by @wighawag is used to create Clones. Each `GameType` has a corresponding implementation within the factory,
and when a new game is created, the factory creates a clone of the `GameType`'s

> implementation contract.

Reviewer 3: What do you think about making it clear the implementation's are already deployed. Just like:

```suggestion
already deployed implementation contract.
```

    /// @dev Owner Permissions:
    ///      - Update the implementation contracts for a given game type.
    /// @return _owner The owner of the contract.
    function owner() public view returns (address _owner);

    /// @notice Creates a new DisputeGame proxy contract.
    /// @param gameType The type of the DisputeGame - used to decide the proxy implementation
    /// @param rootClaim The root claim of the DisputeGame.
    /// @param extraData Any extra data that should be provided to the created dispute game.
    function create(GameType gameType, Claim rootClaim, bytes calldata extraData) external returns (IDisputeGame proxy);

    /// @notice Sets the implementation contract for a specific `GameType`
    /// @param gameType The type of the DisputeGame
    /// @param impl The implementation contract for the given `GameType`
    function setImplementation(GameType gameType, IDisputeGame impl) external;
}
```

## `DisputeGame` Interface

The dispute game interface should be generic enough to allow it to work with any
proof system. This means that it should work fault proofs, validity proofs,
an attestation based proof system, or any other source of truth that adheres to

> the interface.

Reviewer 3: Nicely said 👍

> Clones of the `IDisputeGame`'s `initialize` functions will be called by the `DisputeGameFactory` upon creation.

Reviewer 3: I think this is redundant from line 77-78?

```
When the `DisputeGameFactory` creates a new `DisputeGame` contract, it calls
`initialize()` on the clone to set up the game.
```

Author: It is- felt that the redundancy was worth while here, but glad to remove it if you think that it's been made clear enough in the first statement.

Reviewer 3: Na, happy leaving. It is an important note, and as mark noted, we'll review + refine these specs multiple times for small corrections/nits.

```solidity
////////////////////////////////////////////////////////////////
//                    GENERIC DISPUTE GAME                    //
////////////////////////////////////////////////////////////////

/// @title IDisputeGame
/// @notice The generic interface for a DisputeGame contract.
interface IDisputeGame {
    /// @notice Initializes the DisputeGame contract.
    /// @custom:invariant The `initialize` function may only be called once.
    function initialize() external;

    /// @notice Returns the semantic version of the DisputeGame contract
    function version() external pure returns (string memory _version);

    /// @notice Returns the timestamp that the DisputeGame contract was created at.
    function createdAt() external pure returns (Timestamp _createdAt);

    /// @notice Returns the current status of the game.
    function status() external view returns (GameStatus _status);

>     /// @notice `clones-with-immutable-args` argument #1
>     /// @return _gameType The type of proof system being used.

Reviewer 3: Minor nit, but this seems like an internal detail that should be behind a dev attribute.

```suggestion
    /// @return _gameType The type of proof system being used.
    /// @dev `clones-with-immutable-args` argument #1
```

    /// @dev The reference impl should be entirely different depending on the type (fault, validity)
    ///      i.e. The game type should indicate the security model.
    function gameType() external view returns (GameType _gameType);

>     /// @notice `clones-with-immutable-args` argument #2

Reviewer 3: ibid

    /// @notice `clones-with-immutable-args` argument #3
    /// @return _extraData Any extra data supplied to the dispute game contract by the creator.
    function extraData() external view returns (bytes memory _extraData);

    /// @notice Returns the address of the `BondManager` used 
    function bondManager() public view returns (IBondManager _bondManager);

    /// @notice If all necessary information has been gathered, this function should mark the game
    ///         status as either `CHALLENGER_WINS` or `DEFENDER_WINS` and return the status of
    ///         the resolved game. It is at this stage that the bonds should be awarded to the
    ///         necessary parties.
    /// @dev May only be called if the `status` is `IN_PROGRESS`.
    function resolve() public returns (GameStatus _status);
}

////////////////////////////////////////////////////////////////
//              OUTPUT ATTESTATION DISPUTE GAME               //
////////////////////////////////////////////////////////////////

/// @title IDisputeGame_OutputAttestation
/// @notice The interface for an attestation-based DisputeGame meant to contest output

> ///         proposals in Optimism's `L2OutputOracle` contract.

Reviewer 3: The `L2OutputOracle` will likely be suffixed with `V2`. So, can just do this

```suggestion
///         proposals in Optimism's `L2OutputOracleV2` contract.
```

Author: I believe that once the `L2OutputOracleV2` changes are done, we'll just merge it into the `L2OutputOracle` contract and update Semver.

Reviewer 3: Yea that's cool by me :)

interface IDisputeGame_OutputAttestation is IDisputeGame {
    /// @notice A mapping of addresses from the `signerSet` to booleans signifying whether
    ///         or not they have authorized the `rootClaim` to be invalidated.
    function challenges(address challenger) external view returns (bool _challenged);

>     /// @notice Returns the list of authorized public keys that may challenge the `rootClaim`.
>     /// @return An array of authorized signers that may challenge the `rootClaim`.

Reviewer 3: Seems redundant, consider just doing

```suggestion
    /// @notice The signer set are authorized public keys that may challenge the `rootClaim`.
    /// @return An array of authorized signers.
```

    function signerSet() external view returns (address[] memory _signers);

    /// @notice The amount of signatures required to successfully challenge the `rootClaim`
    ///         output proposal. Once this threshold is met by members of the `signerSet`
    ///         calling `challenge`, the game will be resolved to `CHALLENGER_WINS`.
    /// @custom:invariant The `signatureThreshold` may never be greater than the length of the `signerSet`.
    function signatureThreshold() public view returns (uint16 _signatureThreshold);

    /// @notice Challenge the `rootClaim`.
    /// @dev - If the `ecrecover`ed address that created the signature is not a part of the
    ///      signer set returned by `signerSet`, this function will revert.
    ///      - If the `ecrecover`ed address that created the signature is not the msg.sender,

>     ///      this function will revert.

Reviewer 3: Since this is an interface, what do you think about just saying `should` as opposed to `will`?

I only lean towards `should` since maybe there will be other schemes that unpack the signature differently (?).
